core/src/main/java/io/questdb/network/JavaTlsClientSocket.java

```diff
@@ -0,0 +1,561 @@
+                    sslEngine.wrap(wrapInputBuffer, wrapOutputBuffer);
+                    while (wantsTlsWrite()) {
+                        int n = tlsIO(Socket.WRITE_FLAG);
+                        if (n < 0) {
+                            log.debug().$("could not send TLS close_notify").$();
+                            break;
+                        }
+                    }
+                } catch (SSLException e) {
+                    log.debug().$("could not send TLS close_notify").$(e).$();
+                }
+                sslEngine = null;
+            } // fall through
+            case STATE_PLAINTEXT:
+                state = STATE_CLOSING;
+                // it could be that we allocated buffers but failed to start a TLS session
+                // so we need to free the buffers even in the STATE_PLAINTEXT state
+                freeInternalBuffers();
+                delegate.close();
+                state = STATE_EMPTY;
+                break;
+        }
+    }
+
+    @Override
+    public int getFd() {
+        return delegate.getFd();
+    }
+
+    @Override
+    public boolean isClosed() {
+        return delegate.isClosed();
+    }
+
+    @Override
+    public boolean isTlsSessionStarted() {
+        return sslEngine != null;
+    }
+
+    @Override
+    public void of(int fd) {
+        assert state == STATE_EMPTY;
+        delegate.of(fd);
+        state = STATE_PLAINTEXT;
+    }
+
+    @Override
+    public int recv(long bufferPtr, int bufferLen) {
+        assert sslEngine != null;
+
+        resetBufferToPointer(unwrapOutputBuffer, bufferPtr, bufferLen);
+        unwrapOutputBuffer.position(0);
+
+        try {
+            int plainBytesReceived = 0;
+            for (; ; ) {
+                int n = readFromSocket();
+                assert unwrapInputBuffer.position() == 0 : "unwrapInputBuffer is not compacted";
+                int bytesAvailable = unwrapInputBuffer.limit();
+                if (n < 0 && bytesAvailable == 0) {
+                    if (plainBytesReceived == 0) {
+                        // we didn't manage to read anything from the socket, let's return the error
+                        return n;
+                    }
+                    // we have some data to return, let's return it
+                    return plainBytesReceived;
+                }
+
+
+                if (bytesAvailable == 0) {
+                    // nothing to unwrap, we are done
+                    return plainBytesReceived;
+                }
+
+                SSLEngineResult result = sslEngine.unwrap(unwrapInputBuffer, unwrapOutputBuffer);
+                plainBytesReceived += result.bytesProduced();
+
+                // compact the TLS buffer
+                int bytesConsumed = result.bytesConsumed();
+                int bytesRemaining = bytesAvailable - bytesConsumed;
+                Vect.memcpy(unwrapInputBufferPtr, unwrapInputBufferPtr + bytesConsumed, bytesRemaining);
```

Comment:
I'm not sure if that's a good idea to use memcpy for potentially overlapping memory.

https://en.cppreference.com/w/cpp/string/byte/memcpy says:
> If the objects overlap, the behavior is undefined.
> If the objects are potentially-overlapping or not TriviallyCopyable, the behavior of memcpy is not specified and may be undefined.

`memmove` should be used for overlapping memory.

WDYT?

>

Reply:
that's a good one.!I wasn't sure about this, but I saw it was used like this in other places. let me have a look again. if this is the case then I guess I'm going to fix the other places too.